content/post/formatting-with-style.md

```diff
@@ -0,0 +1,256 @@
+  suggestions of the model, it should be possible to correct its
+  suggestions
+
+To satisfy all those needs with a single approach, we turned to good
+old battle-tested machine learning methods: language models and
+decision trees.
+
+### Language models
+
+In order to define the style of a repository, we turned to the
+well-known Natural Language Processing subjects: language models. A
+language model tries to find out which word most likely follows given
+the sequence of previous words in a sentence. In our case, we'll use
+them to find out which next token is the most likely given the
+previous tokens in some code.
+
+Consider this example where we try to find out what `_` could be:
+
+    if a_
+
+Given the previous tokens `if`, ` ` and `a`, what would be the most
+likely next token `_`? There is not a single correct answer but rather
+a probability distribution over possible tokens. We can learn, for
+example:
+
+    p(" ") = 0.4
+    p(":") = 0.6
+
+`p(" ")` would model the cases where the `if` condition is complex
+(`if a and len(some_list) > 5:`) and `p(":")` would model the cases
+where the `if` condition is already complete (`if a:`).
+
+A language model is simply the set of such probability distributions
+computed for all interesting inputs (all interesting sequences of
+previous tokens).
+
+This model is rather simple and has a big advantage: it doesn't
+require any labeled training data. That means that the engineers who
+will use the model won't have to spend time giving examples of correct
+style and bad style: we'll learn them without any intervention!
+
+To give more power to our language models, we do not restrict what we
+base our predictions on to preceeding tokens but we also inspect the
+words that follow the token that we are predicting. Still, the idea is
+very close to a vanilla language model.
+
+Before diving into the specifics of the models we used, let's
+elaborate on this point and review how we create the window over
+tokens to consider and how we model the elements in those windows.
+
+### Feature Extraction
+
+The backbone of our features is
+[Babelfish](https://docs.sourced.tech/babelfish) (as often in the
+Machine Learning team at source{d}). It allows us to extract Abstract
+Syntax Trees (ASTs) that are universal (UASTs), ie, not language
+dependent.
+
+The advantage of using UASTs over the sequence of tokens is that we
+leverage the structural information: are we in a function? In a
+boolean expression? Is this a lambda expression? We can answer all
+those questions easily with UASTs, while our models would need to work
+extra-hard to understand the structure by itself from token sequences.
+
+Once we have the UAST for a given file, we investigate neighbor tokens
+within a fixed size window around each formatting element we want to
+predict. The shape of the default window is:
+
+- 5 tokens to the left
+- 5 tokens to the right
+- 2 nodes up the parent hierarchy
```

Comment:
up "from"? up "towards"?

Reply:
I have a hard time choosing because the current formulation seems the most fluent to me (and google ngram seems to concur). I'll let it as is for now but please comment to specify which option is the best if needed.